// @shopify/cli-kit/dist/private/node/ui/components/Banner.js
import { Box, Text, useStdout } from 'ink';
import React from 'react';
function typeToColor(type) {
    return {
        success: 'green',
        error: 'red',
        warning: 'yellow',
        info: 'dim',
    }[type];
}
const BANNER_MIN_WIDTH = 80;
function calculateWidth(stdout) {
    const fullWidth = stdout?.columns ?? BANNER_MIN_WIDTH;
    const twoThirdsOfWidth = Math.floor((fullWidth / 3) * 2);
    let width;
    if (fullWidth <= BANNER_MIN_WIDTH) {
        width = fullWidth;
    }
    else if (twoThirdsOfWidth < BANNER_MIN_WIDTH) {
        width = BANNER_MIN_WIDTH;
    }
    else {
        width = twoThirdsOfWidth;
    }
    return width;
}
const Banner = ({ type, children }) => {
    const { stdout } = useStdout();
    return (React.createElement(Box, { width: calculateWidth(stdout), paddingY: 1, paddingX: 2, borderStyle: "round", flexDirection: "column", borderColor: typeToColor(type) },
        React.createElement(Box, { marginTop: -2, marginBottom: 1, marginLeft: -1 },
            React.createElement(Text, { dimColor: true, bold: true }, ` ${type} `)),
        children));
};
export { Banner };
